In [1]:
import numpy as np
import pandas as pd

In [2]:
patients = pd.read_csv("patients.csv")
treatment = pd.read_csv("treatments.csv")
advreact = pd.read_csv("adverse_reactions.csv")
treatcut = pd.read_csv("treatments_cut.csv")

In [3]:
# Copying Datasets
patients_copy = patients.copy()
treatment_copy = treatment.copy()
advreact_copy = advreact.copy()
treatcut_copy = treatcut.copy()

In [4]:
print(patients.shape)
print(treatment.shape)
print(advreact.shape)
print(treatcut.shape)

(503, 14)
(280, 7)
(34, 3)
(70, 7)


In [5]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [6]:
patients.sample(50)

patient_id assigned_sex  given_name      surname  \
360         361         male      Kerman   Dandonneau   
480         481         male      Nasser      Mansour   
184         185       female        Ásta  Grímsdóttir   
57           58       female   Siebrigje    Koldenhof   
238         239         male       Aksel  Vestergaard   
451         452       female      Jia Li         Teng   
189         190       female     Kisanet     Selassie   
15           16         male       Søren         Lund   
340         341       female       Minea     Lindgren   
104         105       female      Yasmin        Silva   
247         248         male      Tuukka   Leppäluoto   
268         269       female       Päivi      Mattila   
391         392         male       Daimy        Tromp   
159         160       female     Coralie      Allaire   
226         227         male        Kári   Hervinsson   
197         198         male         Thể        Lương   
279         280       female  Hanna-Mari      Filatov   
185         186         male       Brice      Authier   
47           48       female         Ana      Correia   
292         293         male      Fraser       Hunter   
383         384         male      Torben    Mortensen   
254         255         male        Kang          Mai   
402         403       female     Manouck      Wubbels   
327         328         male       Torfi    Írisarson   
278         279       female    Generosa        Cabán   
154         155       female        Maya     Isaksson   
97           98         male     Patrick      Gersten   
333         334         male   Jad Allah        Gaber   
344         345       female      Sophia       Haugen   
221         222         male       Csaba         Sági   
66           67         male  Konstantyn    Kowalczyk   
231         232       female    Stefanie       Herman   
209         210       female      Lalita  Eldarkhanov   
410         411         male      Nathan     Cumpston   
431         432       female     Manuela      Cindrić   
361         362         male  Hans Peter     Radislav   
30           31       female      Janina     Bazarova   
239         240         male      Khalid     Johnsrud   
0             1       female         Zoe      Wellish   
386         387         male     Sargent      Flamand   
186         187       female        Jane      Citizen   
461         462         male      Cannan      Cabrera   
408         409       female     Giovana        Rocha   
177         178       female      Monika       Lončar   
88           89         male      Marius       Hansen   
204         205         male   Leixandre       Alanis   
364         365       female  Chukwumoge    Ogochukwu   
158         159         male     Chibuzo        Okoli   
299         300       female       Linda        Lundy   
217         218         male  Saidmohmad   Rushisvili   

                       address               city       state  zip_code  \
360           3688 Adonais Way            Atlanta          GA   30303.0   
480         547 Weekley Street        San Antonio          TX   78212.0   
184           1619 Melm Street       Ormond Beach          FL   32174.0   
57            34 Hamill Avenue            La Mesa  California   91941.0   
238   2246 Pheasant Ridge Road       Philadelphia          PA   19139.0   
451        1251 Clarence Court             Upland  California   91786.0   
189           3227 Park Avenue         Sacramento  California   95817.0   
15      2438 Shady Pines Drive          Kingsport          VA   37660.0   
340       2272 Williams Avenue           Palmdale  California   93550.0   
104         108 Griffin Street            Phoenix          AZ   85012.0   
247         1886 Bicetown Road           New York          NY   10011.0   
268          4320 Rardin Drive         San Carlos          CA   94070.0   
391   522 Lamberts Branch Road            Sunrise          FL   33323.0   
159       1066 Goosetown Drive              Syl

In [7]:
treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [8]:
treatment.sample(50)

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
224      cintia        csorba  47u - 50u          -         8.49       8.00   
252       ellen         luman          -  40u - 39u         9.27       8.77   
244       angel         grant  31u - 37u          -         7.93       7.47   
53       nasser       mansour          -  33u - 31u         7.51       7.06   
146  hanna-mari       filatov  55u - 61u          -         7.71       7.30   
258      csilla      herczegh          -  43u - 46u         7.71       7.27   
64      cumelén    miramontes  24u - 32u          -         7.71       7.33   
173    gabriele       saenger  22u - 30u          -         8.35       7.97   
10       joseph           day  29u - 36u          -         7.70       7.19   
8         saber        ménard          -  54u - 54u         8.08       7.70   
71         hugo       collins          -  53u - 55u         8.00       7.53   
139     barbora       vesecká  29u - 45u          -         7.90       7.46   
164   felicijan        bubanj  62u - 75u          -         7.85       7.50   
75    mackenzie         mckay          -  44u - 43u         9.87       9.48   
82         maya      isaksson  33u - 41u          -         7.66       7.17   
0      veronika      jindrová  41u - 48u          -         7.63       7.20   
176       andré        nordin  40u - 47u          -         7.60       7.27   
17         gina          cain          -  36u - 36u         7.88       7.40   
95       jaakko     eskelinen  39u - 45u          -         7.73       7.42   
137       ivona        jakšić          -  41u - 41u         7.98       7.54   
152       lewis          webb  39u - 44u          -         7.85       7.52   
96    abdul-nur           isa          -  54u - 50u         7.98       7.53   
175      marina      glockner  35u - 45u          -         7.88       7.55   
205       najla         touma          -  39u - 44u         7.66       7.31   
49      jackson       addison          -  42u - 42u         7.99       7.51   
117      javier        moquin          -  30u - 32u         8.00       7.59   
249        kang           mai          -  39u - 36u         7.78       7.45   
7         eddie        archer  31u - 38u          -         7.89       7.55   
11    kristiina        hyypiä          -  36u - 38u         7.87       7.49   
264  chukwumoge     ogochukwu          -  41u - 39u         7.95       7.56   
266      ursula         freud  42u - 54u          -         7.75       7.46   
272      steven           roy          -  41u - 43u         7.87       7.43   
45       zdeněk         synek          -  31u - 31u         7.67       7.32   
130      kamila      pecinová          -  54u - 51u         7.77       7.39   
247     johanna        dreher  51u - 59u          -         7.77       7.34   
14       simone  baumgaertner  27u - 37u          -         7.74       7.30   
270        mika    martinsson  34u - 43u          -         7.50       7.17   
153      nicole     zimmerman          -  59u - 56u         8.98       8.57   
183       tegan       johnson          -  34u - 34u         7.79       7.43   
161     rebecca        priest  40u - 42u          -         7.87       7.57   
199     novalie          berg          -  32u - 31u         7.85       7.49   
254        jose         combs          -  39u - 36u         7.89       7.42   
187        león       reynoso          -  38u - 40u         7.59       7.06   
253      albino     schiavone  35u - 43u          -         7.56       7.15   
124        nate        bowden          -  48u - 47u         9.81       9.30   
113        kari   laatikainen  39u - 43u          -         7.50       7.11   
198      amalie   christensen  38u - 45u          -         7.88       7.55   
89     stefanie        herman  44u - 55u          -         7.72       7.30   
188       bouke        glaser          -  26u - 30u         7.72       7.34   
108         mùi         lương          -  52u - 50u    

In [9]:
advreact

given_name     surname           adverse_reaction
0         berta  napolitani  injection site discomfort
1          lena        baer               hypoglycemia
2        joseph         day               hypoglycemia
3        flavia  fiorentino                      cough
4       manouck     wubbels          throat irritation
5       jasmine       sykes               hypoglycemia
6        louise     johnson               hypoglycemia
7       albinca     komavec               hypoglycemia
8           noe      aranda               hypoglycemia
9         sofia   hermansen  injection site discomfort
10        tegan     johnson                   headache
11         abel     yonatan                      cough
12    abdul-nur         isa               hypoglycemia
13         leon      scholz  injection site discomfort
14     gabriele     saenger               hypoglycemia
15       jia li        teng                     nausea
16        jakob    jakobsen               hypoglycemia
17  christopher    woodward                     nausea
18          ole    petersen               hypoglycemia
19       finley    chandler                   headache
20     anenechi       chidi               hypoglycemia
21     miłosław  wiśniewski  injection site discomfort
22        lixue       hsueh  injection site discomfort
23        merci      leroux               hypoglycemia
24         kang         mai  injection site discomfort
25       elliot  richardson               hypoglycemia
26      clinton      miller          throat irritation
27       idalia       moore               hypoglycemia
28       xiuxiu       chang               hypoglycemia
29         alex    crawford               hypoglycemia
30       monika      lončar               hypoglycemia
31       steven         roy                   headache
32      cecilie      nilsen               hypoglycemia
33    krisztina      magyar               hypoglycemia

In [10]:
treatcut.sample(30)

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
27         zoe       wellish          -  33u - 33u         7.71       7.30   
66       jakob      jakobsen          -  28u - 26u         7.96       7.51   
39      maryam      dratchev  37u - 45u          -         9.91       9.46   
48     nebechi     ekechukwu          -  37u - 39u         7.78       7.39   
0        jožka     resanovič  22u - 30u          -         7.56       7.22   
49     satsita     batukayev          -  42u - 42u         7.63       7.25   
10      edelma   villalpando          -  24u - 26u         7.99       7.56   
24         lóa  hrafnsdóttir  22u - 31u          -         7.60       7.09   
22   beatrycze       woźniak          -  26u - 27u         7.54       7.17   
18      flavia    fiorentino  46u - 57u          -         7.95       7.60   
13    bernarda       cindrić  40u - 49u          -         7.89       7.55   
5         meya       sjöberg          -  42u - 42u         7.77       7.42   
6         sara         miles  30u - 35u          -         7.53       7.12   
43  hans peter      radislav  42u - 53u          -         7.60       7.20   
61        ella          lund  38u - 47u          -         7.57       7.21   
4       amanda       ribeiro  36u - 44u          -         7.85       7.47   
23    miłosław    wiśniewski          -  34u - 33u         7.51       7.08   
41    dalmacia        madrid          -  26u - 23u         7.67       7.21   
32       kifle       mustafa  42u - 50u          -         7.80       7.41   
51     bjørnar        nilsen          -  36u - 33u         7.99       7.70   
7      frydryk       adamski  63u - 74u          -         7.75       7.27   
3          thể         lương          -  61u - 64u         7.64       7.22   
19         noe        aranda  26u - 34u          -         7.51       7.17   
29       luboš         pecha          -  30u - 27u         7.79       7.45   
44      jeremy       montagu          -  52u - 52u         7.68       7.36   
34       adlan      shishani          -  43u - 40u         7.84       7.37   
54       borna     šlezinger          -  42u - 41u         7.55       7.18   
42       daimy         tromp          -  40u - 45u         9.41       8.94   
25   siebrigje     koldenhof  30u - 38u          -         7.90       7.59   
12      annika         vaara          -  20u - 21u         7.73       7.34   

    hba1c_change  
27          0.91  
66          0.95  
39           NaN  
48          0.39  
0           0.34  
49          0.38  
10          0.93  
24           NaN  
22          0.37  
18          0.35  
13          0.34  
5           0.35  
6            NaN  
43           NaN  
61          0.36  
4           0.38  
23          0.93  
41          0.96  
32          0.39  
51           NaN  
7            NaN  
3           0.92  
19          0.34  
29          0.34  
44           NaN  
34          0.97  
54           NaN  
42          0.97  
25          0.31  
12          0.39

In [11]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [12]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [13]:
advreact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   given_name        34 non-null     object
 1   surname           34 non-null     object
 2   adverse_reaction  34 non-null     object
dtypes: object(3)
memory usage: 944.0+ bytes


In [14]:
treatcut.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70 entries, 0 to 69
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    70 non-null     object 
 1   surname       70 non-null     object 
 2   auralin       70 non-null     object 
 3   novodra       70 non-null     object 
 4   hba1c_start   70 non-null     float64
 5   hba1c_end     70 non-null     float64
 6   hba1c_change  42 non-null     float64
dtypes: float64(3), object(4)
memory usage: 4.0+ KB


In [15]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [16]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [17]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [18]:
treatcut.describe()

hba1c_start  hba1c_end  hba1c_change
count    70.000000  70.000000     42.000000
mean      7.838000   7.443143      0.518810
std       0.423007   0.418706      0.270719
min       7.510000   7.020000      0.280000
25%       7.640000   7.232500      0.340000
50%       7.730000   7.345000      0.370000
75%       7.860000   7.467500      0.907500
max       9.910000   9.460000      0.970000

# Quality Issue

[Completeness, Validity, Accuracy, Consistency]

## `Patients`

- Zip code column is stored in decimal format [V]
- Certain Zip codes of 4 digits [V]
- For the state column sometimes full names are use and other times short forms are used [IC]
- patient_id does not consider our algorithm. So we drop it easily.
- Missing values for contact details of 12 patients [C]
- Incorrect data types for assigned_sex and birthdate columns.
- Errorneous data for weights, hights and bmi [Accuracy]
- Inconsistent way of storing phone numbers in contact column [IC]
- Duplicate and not useful data in dataset [IC]

## `Treatment`

- hba1c_change has missing values [C]
- Novodra and Auralin columns has u attached to numbers [V]
- Only 280 rows are there, instead of 350 <u>(Extra 70 rows are in `Treatcut`Dataset)</u> [C]
- Name and surname column is in lower case [IC]
- Incorrect datatype for auralin and novodra columns [V]
- hba1c_change 9 instead of 4 [Accuracy]
- use of dashes instead of NAN in aurolin and novodra columns [IC]

## `Advreact`

- Name and surname column is in lower case [IC]

## `Treatcut`

- hba1c_change has missing values [C]
- Incorrect datatype for auralin and novodra columns [V]
- Novodra and Auralin columns has u attached to numbers [V]
- hba1c_change 9 instead of 4 [Accuracy]
- use of dashes instead of NAN in aurolin and novodra columns [IC]


# Tidiness Issue

## `Patients`

- Phone and email together in one column i.e. contact

## `Treatment`

- 3 variables stored in auralin and novodra column

## `Advreact`

- No point of using this table

## `Treatcut`

- 3 variables stored in auralin and novodra column

# a. Completeness Issues
- Missing values for contact details of 12 patients
- hba1c_change has missing values
- Only 280 rows are there, instead of 350 (Extra 70 rows are in TreatcutDataset)

<b>Define</b>

Missing values for contact details of 12 patients. so drop the rows

<b>Code</b>

In [19]:
patients.dropna(inplace=True)

<b>Test</b>

In [20]:
patients.isnull().sum()

patient_id      0
assigned_sex    0
given_name      0
surname         0
address         0
city            0
state           0
zip_code        0
country         0
contact         0
birthdate       0
weight          0
height          0
bmi             0
dtype: int64

<b>Define</b>
    
Make treatment dataset's rows complete. (treatment+treatcut)

<b>Code</b>

In [21]:
treatment = pd.concat([treatment,treatcut])

<b>Test</b>

In [22]:
treatment.shape

(350, 7)

<b>Define</b>

hba1c_change has missing values. so fill that

<b>Code</b>

In [23]:
treatment['hba1c_change'] = treatment['hba1c_start']-treatment['hba1c_end']

<B>Test</B>

In [24]:
treatment.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0          0.43  
1          0.47  
2          0.43  
3          0.35  
4          0.32

# b. Tidiness Issues

- Phone and email together in one column i.e. contact
- 3 variables stored in auralin and novodra column
- Merge the advreact table to treatment table

<b>Define</b>

Phone and email together in one column i.e. contact. so split it.

<b>Code</b>

In [25]:
patients['email'] = patients['contact'].str.findall('[a-zA-Z_.]+@[a-zA-Z0-9-]+\.[a-zA-Z.]+')

In [26]:
patients['phone']=patients['contact'].str.findall(r'[\+\(]?[1-9][0-9 .\-\(\)]{8,}[0-9]')

In [27]:
patients.drop(['contact'],axis=1,inplace=True)

<b>Test</b>

In [28]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state  zip_code  \
0         576 Brown Bear Drive  Rancho California  California   92390.0   
1    2370 University Hill Road          Armstrong    Illinois   61812.0   
2        1493 Poling Farm Road               York    Nebraska   68467.0   
3          2335 Webster Street         Woodbridge          NJ    7095.0   
4         1428 Turkey Pen Lane             Dothan          AL   36303.0   
..                         ...                ...         ...       ...   
498        2530 Victoria Court       Milton Mills          ME    3852.0   
499   494 Clarksburg Park Road             Sedona          AZ   86341.0   
500          649 Nutter Street      Overland Park          MO   64110.0   
501   3652 Boone Crockett Lane            Seattle          WA   98109.0   
502          2778 North Avenue               Burr    Nebraska   68324.0   

           country  birthdate  weight  height   bmi  \
0    United States  7/10/1976   121.7      66  19.6   
1    United States   4/3/1967   118.8      66  19.2   
2    United States  2/19/1980   177.8      71  24.8   
3    United States  7/26/1951   220.9      70  31.7   
4    United States  2/18/1928   192.3      27  26.1   
..             ...        ...     ...     ...   ...   
498  United States  4/10/1959   181.1      72  24.6   
499  United States  3/26/1948   239.6      70  34.4   
500  United States  1/13/1971   171.2      67  26.8   
501  United States  2/13/1952   176.9      67  27.7   
502  United States   5/3/1954   138.2      71  19.3   

                                   email                phone  
0             [ZoeWellish@superrito.com]       [951-719-9170]  
1                 [PamelaSHill@cuvox.de]  [+1 (217) 569-3204]  
2                 [JaeMDebord@gustr.com]       [402-363-6804]  
3            [PhanBaLiem@jourrapide.com]  [+1 (732) 636-8246]  
4                  [TimNeudorf@cuvox.de]       [334-515-7487]  
..                                   ...                  ...  
498    [MustafaLindstrom@jourrapide.com]       [207-477-0579]  
499             [RumanBisliev@gustr.com]       [928-284-4492]  
500          [JinkedeKeizer@teleworm.us]       [816-223-6007]  
501  [ChidaluOnyekaozulu@jourrapide.com]     [1 360 443 2060]  
502           [PatrickGersten@rhyta.com]       [402-848-4923]  

[491 rows x 15 columns]

<b>Define</b>

3 variables stored in auralin and novodra column. so splite it.

<b>Code</b>

In [29]:
treatment = pd.melt(treatment,id_vars=['given_name','surname','hba1c_start','hba1c_end','hba1c_change'],var_name='treatment',value_name='dose')

In [30]:
treatment = treatment[treatment['dose']!='-']

In [31]:
treatment['low'],treatment['high'] = treatment['dose'].str.split('-').str

C:\anaconda\envs\PythonCPU\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Columnar iteration over characters will be deprecated in future releases.
  """Entry point for launching an IPython kernel.


In [32]:
treatment.drop(columns=['dose'],inplace=True)

<b>Test</b>

In [33]:
treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

    low  high  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

<b>Define</b>

Merge the advreact table to treatment table

<b>Code</b>

In [34]:
treatment = pd.merge(treatment, advreact, how="left", on=['given_name','surname'])

In [35]:
treatment.isnull().sum()

given_name            0
surname               0
hba1c_start           0
hba1c_end             0
hba1c_change          0
treatment             0
low                   0
high                  0
adverse_reaction    315
dtype: int64

In [36]:
treatment['adverse_reaction']=treatment['adverse_reaction'].fillna('No')

In [37]:
def reaction(value):
  if value=='No':
    return 'No'
  else:
    return 'Yes'

In [38]:
treatment['adverse_reaction']=treatment['adverse_reaction'].apply(reaction)

<b>Test</b>

In [39]:
treatment.sample(10)

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
233       sofia     karlsen         7.62       7.15          0.47   novodra   
148      flavia  fiorentino         7.95       7.60          0.35   auralin   
196      mahmud     kadyrov         7.53       7.11          0.42   novodra   
341     timothy      cotton         7.92       7.52          0.40   novodra   
163      regolo       nucci         7.53       7.02          0.51   auralin   
73       yunadi    barsukov         9.47       9.05          0.42   auralin   
162        anja     mueller         7.58       7.29          0.29   auralin   
183  addolorata    lombardi         7.75       7.33          0.42   novodra   
60         urso      aranda         7.60       7.16          0.44   auralin   
314      hiromu    horikawa         7.77       7.28          0.49   novodra   

      low  high adverse_reaction  
233  35u    29u               No  
148  46u    57u              Yes  
196  44u    43u               No  
341  26u    25u               No  
163  51u    59u               No  
73   48u    58u               No  
162  43u    56u               No  
183  49u    46u               No  
60   30u    38u               No  
314  47u    46u               No

# c. Validity Issues
- Zip code column is stored in decimal format
- Certain Zip codes of 4 digits
- Novodra and Auralin columns has u attached to numbers
- Incorrect datatype for auralin and novodra columns

<b>Define</b>

Zip code column is stored in decimal format. it should be intiger.

<b>Code</b>

In [40]:
patients['zip_code'] = patients['zip_code'].astype('int64')

<b>Test</b>

In [41]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 491 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    491 non-null    int64  
 1   assigned_sex  491 non-null    object 
 2   given_name    491 non-null    object 
 3   surname       491 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    int64  
 8   country       491 non-null    object 
 9   birthdate     491 non-null    object 
 10  weight        491 non-null    float64
 11  height        491 non-null    int64  
 12  bmi           491 non-null    float64
 13  email         491 non-null    object 
 14  phone         491 non-null    object 
dtypes: float64(2), int64(3), object(10)
memory usage: 61.4+ KB


<b>Define</b>

Certain Zip codes of 4 digits. it should be five digit. (assuming the last digit is 0)

<b>Code</b>

In [42]:
patients['zip_code']=patients['zip_code'].astype('str').str.pad(5,'right','0')

<b>Test</b>

In [43]:
patients

patient_id assigned_sex given_name      surname  \
0             1       female        Zoe      Wellish   
1             2       female     Pamela         Hill   
2             3         male        Jae       Debord   
3             4         male       Liêm         Phan   
4             5         male        Tim      Neudorf   
..          ...          ...        ...          ...   
498         499         male    Mustafa    Lindström   
499         500         male      Ruman      Bisliev   
500         501       female      Jinke    de Keizer   
501         502       female    Chidalu  Onyekaozulu   
502         503         male        Pat      Gersten   

                       address               city       state zip_code  \
0         576 Brown Bear Drive  Rancho California  California    92390   
1    2370 University Hill Road          Armstrong    Illinois    61812   
2        1493 Poling Farm Road               York    Nebraska    68467   
3          2335 Webster Street         Woodbridge          NJ    70950   
4         1428 Turkey Pen Lane             Dothan          AL    36303   
..                         ...                ...         ...      ...   
498        2530 Victoria Court       Milton Mills          ME    38520   
499   494 Clarksburg Park Road             Sedona          AZ    86341   
500          649 Nutter Street      Overland Park          MO    64110   
501   3652 Boone Crockett Lane            Seattle          WA    98109   
502          2778 North Avenue               Burr    Nebraska    68324   

           country  birthdate  weight  height   bmi  \
0    United States  7/10/1976   121.7      66  19.6   
1    United States   4/3/1967   118.8      66  19.2   
2    United States  2/19/1980   177.8      71  24.8   
3    United States  7/26/1951   220.9      70  31.7   
4    United States  2/18/1928   192.3      27  26.1   
..             ...        ...     ...     ...   ...   
498  United States  4/10/1959   181.1      72  24.6   
499  United States  3/26/1948   239.6      70  34.4   
500  United States  1/13/1971   171.2      67  26.8   
501  United States  2/13/1952   176.9      67  27.7   
502  United States   5/3/1954   138.2      71  19.3   

                                   email                phone  
0             [ZoeWellish@superrito.com]       [951-719-9170]  
1                 [PamelaSHill@cuvox.de]  [+1 (217) 569-3204]  
2                 [JaeMDebord@gustr.com]       [402-363-6804]  
3            [PhanBaLiem@jourrapide.com]  [+1 (732) 636-8246]  
4                  [TimNeudorf@cuvox.de]       [334-515-7487]  
..                                   ...                  ...  
498    [MustafaLindstrom@jourrapide.com]       [207-477-0579]  
499             [RumanBisliev@gustr.com]       [928-284-4492]  
500          [JinkedeKeizer@teleworm.us]       [816-223-6007]  
501  [ChidaluOnyekaozulu@jourrapide.com]     [1 360 443 2060]  
502           [PatrickGersten@rhyta.com]       [402-848-4923]  

[491 rows x 15 columns]

<b>Define</b>

Novodra and Auralin columns has u attached to numbers. (Now it becomes low and high)

<b>Code</b>

In [44]:
treatment['low']=treatment['low'].str.split('u').str[0]

In [45]:
treatment['high']=treatment['high'].str.split('u').str[0]

<b>Test</b>

In [46]:
treatment.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment low  \
0   veronika    jindrová         7.63       7.20          0.43   auralin  41   
1       skye  gormanston         7.97       7.62          0.35   auralin  33   
2     sophia      haugen         7.65       7.27          0.38   auralin  37   
3      eddie      archer         7.89       7.55          0.34   auralin  31   
4       asia     woźniak         7.76       7.37          0.39   auralin  30   

  high adverse_reaction  
0   48               No  
1   36               No  
2   42               No  
3   38               No  
4   36               No

<b>Define</b>

Incorrect datatype for auralin and novodra columns

<b>Code</b>

In [47]:
treatment['low']=treatment['low'].astype('int32')
treatment['high']=treatment['high'].astype('int32')

<b>Test</b>

In [48]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 350 entries, 0 to 349
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   given_name        350 non-null    object 
 1   surname           350 non-null    object 
 2   hba1c_start       350 non-null    float64
 3   hba1c_end         350 non-null    float64
 4   hba1c_change      350 non-null    float64
 5   treatment         350 non-null    object 
 6   low               350 non-null    int32  
 7   high              350 non-null    int32  
 8   adverse_reaction  350 non-null    object 
dtypes: float64(3), int32(2), object(4)
memory usage: 24.6+ KB


# d. Accuracy Issues
- Errorneous data for weights and hights

<b>Define</b>

Errorneous data for weights, hights, bmi<br>
(bmi=703 x ((WEIGHT)/(HEIGHT)^2))

<b>Code</b>

In [49]:
patients.describe()

patient_id      weight      height         bmi
count  491.000000  491.000000  491.000000  491.000000
mean   251.957230  173.346640   66.643585   27.460285
std    147.060588   33.912482    4.425028    5.277615
min      1.000000   48.800000   27.000000   17.100000
25%    123.500000  149.300000   63.000000   23.250000
50%    252.000000  175.600000   67.000000   27.200000
75%    380.500000  199.400000   70.000000   31.650000
max    503.000000  255.900000   79.000000   37.700000

- minimum hight = 27, weight = 192.3 and bmi = 26.1<br>
- minimum weight = 48.8, hight = 63 and bmi = 19.1

In [50]:
patients[patients['weight'] == 48.8]

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state zip_code        country   birthdate  weight  height   bmi  \
210  Wooster    OH    44691  United States  11/26/1938    48.8      63  19.1   

                               email           phone  
210  [CamillaZaitseva@superrito.com]  [330-202-2145]

In [51]:
patients[patients['height'] == 27.0]

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state zip_code        country  birthdate  weight  height   bmi  \
4    AL    36303  United States  2/18/1928   192.3      27  26.1   

                   email           phone  
4  [TimNeudorf@cuvox.de]  [334-515-7487]

In [52]:
patients['height']=patients['height'].replace(27,72) # because weight=192.3 and bmi=26.1 applying bmi rule we get hight=72

In [53]:
patients['weight']=patients['weight'].replace(48.8,107.8) # apply bmi rule we get weight=107.8

<b>Test</b>

In [54]:
patients.describe()

patient_id      weight      height         bmi
count  491.000000  491.000000  491.000000  491.000000
mean   251.957230  173.466802   66.735234   27.460285
std    147.060588   33.573102    4.052608    5.277615
min      1.000000  102.100000   59.000000   17.100000
25%    123.500000  149.300000   63.000000   23.250000
50%    252.000000  175.600000   67.000000   27.200000
75%    380.500000  199.400000   70.000000   31.650000
max    503.000000  255.900000   79.000000   37.700000

# e. Consistency Issues

- For the state column sometimes full names are use and other times short forms are used
- Duplicate and not useful data in dataset
- Name and surname column is in lower case

<b>Define</b>

For the state column sometimes full names are use and other times short forms are used

<b>Code</b>

In [55]:
patients_copy['state'].value_counts()

California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
OH            14
Illinois      14
MI            13
LA            13
Florida       13
OK            13
NJ            12
VA            11
WI            10
MS            10
IL            10
MN             9
FL             9
IN             9
TN             9
AL             9
KY             8
NC             8
WA             8
MO             7
ID             6
NV             6
KS             6
SC             5
IA             5
CT             5
RI             4
ND             4
AZ             4
CO             4
AR             4
ME             4
Nebraska       4
DE             3
OR             3
MD             3
WV             3
SD             3
MT             2
NE             2
DC             2
VT             2
AK             1
NM             1
WY             1
NH             1
Name: state, dtype: int64

In [56]:
patients['state']=patients['state'].replace('California','CA')
patients['state']=patients['state'].replace('New York','NY')
patients['state']=patients['state'].replace('Illinois','IL')
patients['state']=patients['state'].replace('Florida','FL')
patients['state']=patients['state'].replace('Nebraska','NE')

In [57]:
patients['state'].value_counts()

CA    60
NY    47
TX    32
IL    24
MA    22
FL    22
PA    18
GA    15
OH    14
OK    13
LA    13
MI    13
NJ    12
VA    11
WI    10
MS    10
AL     9
TN     9
IN     9
MN     9
KY     8
NC     8
WA     8
MO     7
KS     6
ID     6
NE     6
NV     6
SC     5
IA     5
CT     5
AR     4
AZ     4
RI     4
ME     4
CO     4
ND     4
DE     3
SD     3
WV     3
MD     3
OR     3
VT     2
MT     2
DC     2
NH     1
NM     1
AK     1
WY     1
Name: state, dtype: int64

<b>Test</b>

In [58]:
patients.sample(10)

patient_id assigned_sex given_name     surname                address  \
417         418       female     Xiuxiu        Yuan  4277 Mutton Town Road   
386         387         male    Sargent     Flamand    163 Hide A Way Road   
120         121       female   Nicoline  Østergaard       2836 Boring Lane   
484         485         male    Placido     Udinesi      1094 Jones Avenue   
395         396         male  Krisztián       Bakos       2316 Ingram Road   
185         186         male      Brice     Authier   1895 Hart Ridge Road   
205         206         male      Breno     Correia       3539 Bottom Lane   
328         329       female       Anja      Hueber   3216 Lodgeville Road   
382         383         male     Diðrik   Maríasson     4015 Juniper Drive   
240         241       female   Marphisa   Compagnon     3391 Marcus Street   

               city state zip_code        country   birthdate  weight  height  \
417         Redmond    WA    98052  United States    3/8/1943   150.0      66   
386        San Jose    CA    95134  United States   5/10/1965   207.7      65   
120   San Francisco    CA    94108  United States  12/14/1926   124.5      63   
484      Greensboro    NC    28716  United States   5/31/1934   175.8      65   
395     Yadkinville    NC    27055  United States   3/28/1975   231.9      74   
185          Elkton    MI    48731  United States  12/16/1939   125.1      64   
205       Tonawanda    NY    14150  United States   7/11/1974   180.6      67   
328     Minneapolis    MN    55402  United States   4/16/1987   151.8      65   
382  Mount Pleasant    MI    48858  United States   3/25/1949   149.6      71   
240      Huntsville    AL    35806  United States    3/9/1967   198.0      63   

      bmi                               email                phone  
417  24.2          [XiuxiuYuan@superrito.com]       [360-689-4989]  
386  34.6         [SargentFlamand@dayrep.com]       [408-215-6012]  
120  22.1  [NicolinePstergaard@superrito.com]       [415-676-8818]  
484  29.3         [PlacidoUdinesi@dayrep.com]       [336-697-2005]  
395  29.8        [BakosKrisztian@teleworm.us]       [336-677-8769]  
185  21.5        [BriceAuthier@superrito.com]     [1 989 375 2216]  
205  28.3    [BrenoLimaCorreia@superrito.com]       [716-743-5884]  
328  25.3            [AnjaHueber@teleworm.us]  [+1 (612) 342-6065]  
382  20.9     [DirikMariasson@jourrapide.com]       [989-822-1351]  
240  35.1        [MarphisaCompagnon@cuvox.de]  [+1 (256) 615-5522]

<b>Define</b>

Duplicate and not useful data in dataset.so drop it

<b>Code</b>

In [59]:
treatment.drop_duplicates(subset=['given_name','surname'],inplace=True)

<b>Test</b>

In [60]:
treatment[treatment['given_name']=='joseph']

given_name surname  hba1c_start  hba1c_end  hba1c_change treatment  low  \
5       joseph     day         7.70       7.19          0.51   auralin   29   
167     joseph  tucker         7.67       7.30          0.37   auralin   48   

     high adverse_reaction  
5      36              Yes  
167    56               No

<b>Define</b>

Name and surname column is in lower case. so fix it.

<b>Code</b>

In [61]:
def capitalize(value):
    return value.title()

In [62]:
patients['given_name']=patients['given_name'].apply(capitalize)
treatment['given_name']=treatment['given_name'].apply(capitalize)
advreact['given_name']=advreact['given_name'].apply(capitalize)
patients['surname']=patients['surname'].apply(capitalize)
treatment['surname']=treatment['surname'].apply(capitalize)
advreact['surname']=advreact['surname'].apply(capitalize)

<b>Test</b>

In [63]:
treatment.sample(3)

given_name   surname  hba1c_start  hba1c_end  hba1c_change treatment  \
345  Christopher  Woodward         7.51       7.06          0.45   novodra   
133       Albina  Zetticci         7.93       7.73          0.20   auralin   
234     Flaviano  Endrizzi         7.93       7.56          0.37   novodra   

     low  high adverse_reaction  
345   55    51              Yes  
133   45    51               No  
234   46    43               No

In [64]:
advreact.sample(3)

given_name     surname           adverse_reaction
0       Berta  Napolitani  injection site discomfort
26    Clinton      Miller          throat irritation
32    Cecilie      Nilsen               hypoglycemia

# f. Other Issues Fixing

In [65]:
patients.drop('patient_id',axis=1,inplace=True)

In [66]:
patients['assigned_sex']=patients['assigned_sex'].astype('category')

In [67]:
patients['birthdate']= pd.to_datetime(patients['birthdate'])

In [68]:
patients['birthdate'] = patients['birthdate'].dt.strftime('%d/%m/%Y')

In [69]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 491 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype   
---  ------        --------------  -----   
 0   assigned_sex  491 non-null    category
 1   given_name    491 non-null    object  
 2   surname       491 non-null    object  
 3   address       491 non-null    object  
 4   city          491 non-null    object  
 5   state         491 non-null    object  
 6   zip_code      491 non-null    object  
 7   country       491 non-null    object  
 8   birthdate     491 non-null    object  
 9   weight        491 non-null    float64 
 10  height        491 non-null    int64   
 11  bmi           491 non-null    float64 
 12  email         491 non-null    object  
 13  phone         491 non-null    object  
dtypes: category(1), float64(2), int64(1), object(10)
memory usage: 54.3+ KB


In [70]:
treatment['adverse_reaction']=treatment['adverse_reaction'].astype('category')

In [71]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 349 entries, 0 to 349
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   given_name        349 non-null    object  
 1   surname           349 non-null    object  
 2   hba1c_start       349 non-null    float64 
 3   hba1c_end         349 non-null    float64 
 4   hba1c_change      349 non-null    float64 
 5   treatment         349 non-null    object  
 6   low               349 non-null    int32   
 7   high              349 non-null    int32   
 8   adverse_reaction  349 non-null    category
dtypes: category(1), float64(3), int32(2), object(3)
memory usage: 22.2+ KB


In [72]:
advreact['adverse_reaction']=advreact['adverse_reaction'].astype('category')

In [73]:
advreact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   given_name        34 non-null     object  
 1   surname           34 non-null     object  
 2   adverse_reaction  34 non-null     category
dtypes: category(1), object(2)
memory usage: 914.0+ bytes
